src/viam/sdk/components/camera.cpp

```diff
@@ -24,6 +29,106 @@ API API::traits<Camera>::api() {
+    std::memcpy(&value, &data[*offset], sizeof(uint64_t));
+    value = boost::endian::big_to_native(value);
+    *offset += sizeof(uint64_t);
+
+    return value;
+}
+
+}  // namespace
+
+std::vector<unsigned char> Camera::encode_depth_map(const Camera::depth_map& m) {
+    const auto& shape = m.shape();
+    if (shape.size() != 2) {
+        throw Exception("Depth map is not a 2D array.");
+    }
+
+    const size_t height = shape[0];
+    const size_t width = shape[1];
+    const size_t total_byte_count = k_header_size + height * width * sizeof(uint16_t);
+    std::vector<unsigned char> data(total_byte_count);
+    size_t offset = 0;
+
+    // Network data is stored in big-endian, while most host systems are little endian.
+    append_uint64_big_endian(data, &offset, k_magic_number);
+    append_uint64_big_endian(data, &offset, width);
+    append_uint64_big_endian(data, &offset, height);
+
+    for (size_t i = 0; i < height; ++i) {
+        for (size_t j = 0; j < width; ++j) {
+            const uint16_t value = m(i, j);
+            const uint16_t value_be = boost::endian::native_to_big(value);
+            std::memcpy(&data[offset], &value_be, sizeof(uint16_t));
+            offset += sizeof(uint16_t);
+        }
+    }
+
+    return data;
+}
+
+Camera::depth_map Camera::decode_depth_map(const std::vector<unsigned char>& data) {
+    if (data.size() < k_header_size) {
+        throw Exception("Data too short to contain valid depth information. Size: " +
+                        std::to_string(data.size()));
+    }
+
+    size_t offset = 0;
+    const uint64_t magic_number = read_uint64_big_endian(data, &offset);
+    if (magic_number != k_magic_number) {
+        throw Exception(
+            "Invalid header for a vnd.viam.dep encoded depth image. The data may be corrupted, or "
+            "is not a Viam-encoded depth map.");
+    }
+
+    const uint64_t width = read_uint64_big_endian(data, &offset);
+    const uint64_t height = read_uint64_big_endian(data, &offset);
+
+    const auto expected_size = k_header_size + width * height * sizeof(uint16_t);
+    if (data.size() != expected_size) {
+        throw Exception("Data size does not match width, height, and depth values. Actual size: " +
+                        std::to_string(data.size()) +
+                        ". Expected size: " + std::to_string(expected_size) +
+                        ". Width: " + std::to_string(width) + " Height: " + std::to_string(height));
+    }
+
+    std::vector<uint16_t> depth_values;
+    depth_values.reserve(width * height);
+    for (size_t i = 0; i < width * height; ++i) {
+        const size_t data_index = k_header_size + i * sizeof(uint16_t);
+        const uint16_t depth_value = static_cast<uint16_t>(
+            data[data_index] << 8 | data[data_index + 1]);  // Assemble from big endian into uint16
+        depth_values.push_back(depth_value);
+    }
+
+    return xt::adapt(depth_values, std::array<size_t, 2>{height, width});
```

Comment:
I think this is going to cause problems: I think `xt::adapt` is only going to view the memory in `depth_values`, and it will go out of scope when this function returns.

I think you will need another approach to construct an xtensor that owns the memory here. Let me know if you need me to look into how to do that best.